gradual type constraint generation

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* #80102
* __->__ #80095


Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node.

migeed-z committed:
gradual type constraint generation

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"




[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"




[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387327).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387327).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387327).

jamesr66a reviewed (commented):
Can you add a description? Would be more effective for me to review if there was a high-level description of the functionality being added

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -0,0 +1,334 @@
+import torch
+import operator
+from typing import Callable, Dict
+from torch.fx.experimental.migrate_gradual_types.constraint import ApplyBroadcasting, CalcProduct  # type: ignore[import]
```
Why all the `type: ignore`? Is lint complaining? Seems like something is broken if so, because `constraint.py` is there, right?

migeed-z replied:
Fixed. The __init__.py file was missing.

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -0,0 +1,334 @@
+from torch.fx.tensor_type import Dyn, TensorType
+from torch.nn.modules.conv import Conv2d
+from torch.nn.modules.batchnorm import BatchNorm2d
+_INFERENCE_RULES: Dict[Target, Callable] = {}
```
nit: add some vertical whitespace before this

migeed-z replied:
added

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -0,0 +1,334 @@
+            raise NotImplementedError(f"Method {n.op} not yet implemented")
+
+
+def generate_flatten_constraints(start_dim, end_dim, input, flattened, n, counter):
```
Can you move this function to the beginning of the file before it's used? I was confused because I hadn't seen it while reading linearly

jamesr66a commented on test/fx/test_z3_gradual_types.py:
```diff
@@ -0,0 +1,51 @@
+        generator = ConstraintGenerator(traced)
+        new_constraints, counter = generator.generate_constraints(0)
+        assert len(new_constraints.conjucts) == 16
+
```
Seems like we don't have coverage for all the constraints that we generate. Is that right? Can we add more test cases?

Also, is there a way to test these constraints against the actual shapes?

migeed-z replied:
For this stage, there aren't any actual shapes. We just generate some constraints. I think the only way to check if it gives us anything meaningful is to transform them to z3 constraints and get actual shapes and compare them against specific inputs. This part is in the next PR and it calls this function as a subroutine.

jamesr66a reviewed (commented):
Looking good!

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"


Differential Revision: [D37387327](https://our.internmc.facebook.com/intern/diff/D37387327)

Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

jamesr66a reviewed (commented):
Looks good to me. Ping @anijain2305 @jansel for review

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -0,0 +1,329 @@
+    # c5 = BinConstraintT(e1, 2, op_leq)
+    # c6 = BinConstraintT(e2, 2, op_leq)
+    # c7 = BinConstraintT(e11, 2, op_leq)
+    # c8 = BinConstraintT(e22, 2, op_leq)
```
removed

migeed-z commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -0,0 +1,329 @@
+    c1 = Conj([input_dyn, output_dyn])
+
+    c2 = []
+    for i in range(1, 5):
```
done

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z committed:
Update on "gradual type constraint generation"



Generate high-level gradual type constraints for an fx graph. Those constraints include operations such as matching and consistency and are generated per node. 

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g
